mne_bids/write.py

```diff
@@ -176,6 +176,28 @@ def _events_tsv(events, raw, fname, trial_type, overwrite=False,
     return fname
 
 
+def _montage_tsv(raw, fname, overwrite=False, verbose=True):
+    from collections import OrderedDict
+    from mne.utils import _check_ch_locs
+    x, y, z, names = list(), list(), list(), list()
+    for ch in raw.info['chs']:
+        if _check_ch_locs([ch]):
+            x.append(ch['loc'][0])
+            y.append(ch['loc'][1])
+            z.append(ch['loc'][2])
+        else:
+            x.append('n/a')
+            y.append('n/a')
+            z.append('n/a')
+        names.append(ch['ch_name'])
+    data = OrderedDict([('name', names),
+                        ('x', x),
+                        ('y', y),
+                        ('z', z)])
+    _write_tsv(fname, data, overwrite=overwrite, verbose=verbose)
+    return fname
```

Comment:
I think this is slightly problematic, because like this, we will write `electrodes.tsv` data that might originate from an idealized `Montage` as opposed to a measured `DigMontage` :thinking: 

for BIDS data, we only want to write electrode positions (x,y,z) if these have been measured/digitized ... not some template or idealized positions

cc @jasmainak 

I'll have to figure out if it's possible in MNE to distinguish between a DigMontage and a Montage ... currently it seems like once they are set, they are treated equally.

Ideally, I would vote for only writing `electrodes.tsv` files if we know that
- the electrode positions are MEASURED
- we have NAS, PRA, LPA to scale the positions to FIFF space

Reply:
why is that? Does the BIDS specification make this mandate somewhere? I couldn't find it with a quick search ...

Reply:
i don't think much consideration has been given toward template data in the digitization in the bids discussion. i think it would be nice to have this information but if the analysis ultimately rest on the template, the digpoint data should exist somewhere (in the electrode file or in the attached template). it would be incumbent on the researcher to describe it in their methods somehow.

Reply:
To me either you have it described in your electrode file, if you have a digitization or in the template if you are using a template. (and the template structure should be a replica of electrode file).

Reply:
I feel like researchers / data-providers should only provide measured data (=digitized, true sensor positions) ... template positions etc. should be provided by the analysis software.

Sharing template positions without an indicator that the positions in question are in fact just some "standard", "idealized" positions can easily lead to misunderstandings, where somebody thinks that these are true positions.

With this reasoning, a researcher would write an `electrodes.tsv` and accompanying `coordsystem.json` only if they actually have true positions ... else, they just omit this info and the analysis software can provide a template based on the data format.

Think about it this way: If researchers / data-providers also shared an `electrodes.tsv` for *template* positions for each dataset and datafile, we'd have a lot of duplication and clutter ... because the template positions are always the same, depending on the sensor name and manufacturer.

Reply:
I would argue that this should be fixed in the BIDS specification not in MNE. If you share template positions, there should be clear indicator in the specification that it's a template. Ultimately, many researchers won't care as long as they can run their pipeline. 

If you share the electrode locations only in some BIDS datasets, how is the researcher supposed to guess which template it belongs to?

Reply:
> I would argue that this should be fixed in the BIDS specification not in MNE. If you share template positions, there should be clear indicator in the specification that it's a template.

I agree

> Ultimately, many researchers won't care as long as they can run their pipeline.

Sounds very sad, but maybe true

> If you share the electrode locations only in some BIDS datasets, how is the researcher supposed to guess which template it belongs to?

Example: For EEG ... if the electrodes have names such as "FC1, Cz, ..." we know that it's the 10-20 system and these positions are perfectly defined on the sphere - I assume this is true for all kinds of sensors (templates can almost always be unambiguously guessed)

Reply:
We should have a live discussion on this with also @massich and @larsoner 

What you say here is that having 2 classes Montage and DigMontage can have a real motivation (template vs real dig). MNE stores chs pos and dig. I could have made sense to just put template locations in chs and not in dig also we have the set_dig parameter in set_montage method.

Anyone has a strong feeling about this?

Reply:
Basically this is an argument for keeping the status quo (different behavior of Montage and DigMontage, different classes, etc.) and just updating it so that the former only updates `info['chs']` (well that adding three fiducials to `info['dig']`) and the latter updates both that and `info['dig']`. This is unfortunate because keeping these two classes around is what we wanted to overhaul/unify. It will be more work and lead to more complex code (i.e., maintenance overhead and bugs) than having one.

If we absolutely have to we probably can unify a lot of the code and behaviors while sticking with two classes. For example `plot_alignment` will now need to do more than triage based on `info['dig']`, it will also need to look if `info['chs']` has EEG channels and if those channels have positions and those positions are not in `info['dig']`.

I'd prefer either (or both):

1. Fixing this at the BIDS end
2. Getting a new FIFF constant for this like we did for `CUSTOM_REF_APPLIED`. Both seem like better solutions than making our code more complex.

In particular for (2) we might be able to make a constant like `FIFF_DIGITIZATION_SOURCE` which could be a str. For most Neuromag folks it would be `'Polhemus'` but could be other things like `'Structure sensor'`, and for our montages we could have a standard that it's `Montage: standard_1020.txt` or something. Then we have a consistent way to indicate the dig source.

Reply:
If it's only reading the locations and setting this `FIFF_DIGITIZATION_SOURCE` as an enumerate of templates + input systems. (i.e:`{'standard_1020', 'standard_1005', ..., 'Polhemus', 'BVCT'}`)

This is fine to have a single class and warn about possible inaccuracies when the data has been read from a template. The problem comes if this is more than just reading then keeping a flag, etc might not be a good solution. 

I agree that we should have a talk about it.

Reply:
> I assume this is true for all kinds of sensors (templates can almost always be unambiguously guessed)

This is not true because if this were true, you would not need `raw.set_montage` method or `montage` argument in the raw readers. As we are practical people, I invite you to try writing code to unambiguously guess it ;-) If it cannot be guessed in 5 tries, then the burden of sharing the electrode locations should lie on the researcher sharing the data, no?

Reply:
> If it cannot be guessed in 5 tries, then the burden of sharing the electrode locations should lie on the researcher sharing the data, no?

okay, fair enough --> I already agreed that this could/should be fixed in BIDS (by adding a key-vaul pair whether the electrode locations are template positions or digitized positions)

but not everybody is using BIDS, so we will also have to deal with this in MNE-Python. The suggestion by @larsoner seems good to me at a first glance (using BOTH his points)

Reply:
just fyi, this is what Robert Oostenveld thinks about this: https://github.com/bids-standard/pyedf/issues/7#issuecomment-524061032

... the thread might be an interesting read for everybody working on digitization in MNE-Python

Reply:
For me BIDS (MEG, EEG and iEEG) so far have always been on raw (i.e. measured and unprocessed) data, not on processed (template) or model data (like electrodes on a sphere). That might have had to be stressed more. 

> I would argue that this should be fixed in the BIDS specification not in MNE. If you share template positions, there should be clear indicator in the specification that it's a template. Ultimately, many researchers won't care as long as they can run their pipeline.

Agreed, I propose that the BIDS spec for raw data is updated to clarify that people should include measured data, not templates. Should that also be extended to template anatomical MRIs? I.e. would people right now also be including for example colin27 as anat/sub-01_T1w.nii, in case they don't have an anatomical MRI?

> If you share the electrode locations only in some BIDS datasets, how is the researcher supposed to guess which template it belongs to?

That has been given explicit thought and is part of the specification: you [SHOULD specify](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/03-electroencephalography.html#sidecar-json-_eegjson) CapManufacturersModelName and EEGPlacementScheme. These are more precise than merely relying on electrode names, since cap designs vary between manufacturers and sometimes between Caucasian/Asian fit. If you (as experimenter) tell me what cap you used - I (as software developer) can provide you with the template - which I can also update and improve over the years. E.g. I might start with a [spherical](https://robertoostenveld.nl/electrode/) 10-20 template, then in a later version of the software update that for a [more accurate](https://github.com/fieldtrip/fieldtrip/blob/11c2cc8898be692f94c9cdfd7ccd8d49c3cc00c8/ft_electrodeplacement.m#L23) one.  Ideally EEG cap manufacturers would get themselves a digitizer, scan all of their cap designs and share those (as we are arguing for [here](https://doi.org/10.1016/j.jneumeth.2019.108378)).

Reply:
I opened an issue for this on BIDS-Specification: https://github.com/bids-standard/bids-specification/issues/318

for MNE-BIDS, we "solved" this in our last call, I summarize the solution here: https://github.com/mne-tools/mne-bids/issues/264